File: crates/solobase/src/blocks/system.rs
use wafer_run::block::{Block, BlockInfo};
use wafer_run::context::Context;
use wafer_run::types::*;
use wafer_run::helpers::*;

pub struct SystemBlock;

#[cfg_attr(target_arch = "wasm32", async_trait::async_trait(?Send))]
#[cfg_attr(not(target_arch = "wasm32"), async_trait::async_trait)]
impl Block for SystemBlock {
    fn info(&self) -> BlockInfo {
        BlockInfo {
            name: "@solobase/system".to_string(),
            version: "1.0.0".to_string(),
            interface: "http.handler".to_string(),
            summary: "System health, debug, and navigation endpoints".to_string(),
            instance_mode: InstanceMode::Singleton,
            allowed_modes: vec![InstanceMode::Singleton],
            admin_ui: None,
            runtime: wafer_run::types::BlockRuntime::Native,
            requires: Vec::new(),
        }
    }

    async fn handle(&self, _ctx: &dyn Context, msg: &mut Message) -> Result_ {
        let path = msg.path();

        match path {
            "/health" => {
                let resp = serde_json::json!({"status": "ok"});
                json_respond(msg, &resp)
            }
            "/debug/time" => {
                let now = chrono::Utc::now();
                let resp = serde_json::json!({
                    "utc": now.to_rfc3339(),
                    "unix": now.timestamp(),
                    "unix_ms": now.timestamp_millis()
                });
                json_respond(msg, &resp)
            }
            "/nav" => {
                let nav = serde_json::json!([
                    {"id": "dashboard", "title": "Dashboard", "label": "Dashboard", "href": "/admin", "path": "/admin", "icon": "LayoutDashboard"},
                    {"id": "users", "title": "Users", "label": "Users", "href": "/admin/users", "path": "/admin/users", "icon": "Users"},
                    {"id": "database", "title": "Database", "label": "Database", "href": "/admin/database", "path": "/admin/database", "icon": "Database"},
                    {"id": "iam", "title": "IAM", "label": "IAM", "href": "/admin/iam", "path": "/admin/iam", "icon": "Shield"},
                    {"id": "logs", "title": "Logs", "label": "Logs", "href": "/admin/logs", "path": "/admin/logs", "icon": "FileText"},
                    {"id": "settings", "title": "Settings", "label": "Settings", "href": "/admin/settings", "path": "/admin/settings", "icon": "Settings"},
                    {"id": "legalpages", "title": "Legal Pages", "label": "Legal Pages", "href": "/admin/legalpages", "path": "/admin/legalpages", "icon": "Scale"},
                    {"id": "products", "title": "Products", "label": "Products", "href": "/admin/ext/products", "path": "/admin/ext/products", "icon": "ShoppingBag"},
                    {"id": "files", "title": "Files", "label": "Files", "href": "/admin/storage", "path": "/admin/storage", "icon": "FolderOpen"},
                    {"id": "custom-tables", "title": "Custom Tables", "label": "Custom Tables", "href": "/admin/custom-tables", "path": "/admin/custom-tables", "icon": "Table"}
                ]);
                json_respond(msg, &nav)
            }
            _ => err_not_found(msg, "not found"),
        }
    }

    async fn lifecycle(&self, _ctx: &dyn Context, _event: LifecycleEvent) -> std::result::Result<(), WaferError> {
        Ok(())
    }
}

File: crates/solobase-cloudflare/worker/blocks-ts/system.ts
// TypeScript-native system block handler.

import type { Message, BlockResult, TenantConfig } from '../types';
import type { RuntimeHost } from '../host';
import { metaGet } from '../convert';
import { getUsageSummary } from '../usage';

export function handle(msg: Message, host?: RuntimeHost): BlockResult | Promise<BlockResult> {
  const path = metaGet(msg.meta, 'req.resource') ?? '/';

  if (path === '/health') {
    return jsonRespond(msg, { status: 'ok' });
  }

  if (path === '/debug/time') {
    const now = new Date();
    return jsonRespond(msg, {
      utc: now.toISOString(),
      unix: Math.floor(now.getTime() / 1000),
      unix_ms: now.getTime(),
    });
  }

  if (path === '/nav') {
    return jsonRespond(msg, [
      { id: 'dashboard', title: 'Dashboard', href: '/admin', icon: 'LayoutDashboard' },
      { id: 'users', title: 'Users', href: '/admin/users', icon: 'Users' },
      { id: 'database', title: 'Database', href: '/admin/database', icon: 'Database' },
      { id: 'iam', title: 'IAM', href: '/admin/iam', icon: 'Shield' },
      { id: 'logs', title: 'Logs', href: '/admin/logs', icon: 'FileText' },
      { id: 'settings', title: 'Settings', href: '/admin/settings', icon: 'Settings' },
    ]);
  }

  if (path === '/b/usage') {
    return handleUsage(msg, host);
  }

  return { action: 'error', error: { code: 'not-found', message: 'not found', meta: [] } };
}

async function handleUsage(msg: Message, host?: RuntimeHost): Promise<BlockResult> {
  // Usage is only available for tenant API requests (not platform)
  // We need access to the D1 database via the host
  if (!host) {
    return jsonRespond(msg, { error: 'usage not available' });
  }

  // Get tenant info from dispatch context — passed via meta by the worker
  // For now, return a basic response; the actual usage query needs the DB directly
  // which we'll access through a special service call
  try {
    const result = await host.callBlock('wafer-run/database', {
      kind: 'database.query_raw',
      data: new TextEncoder().encode(JSON.stringify({
        query: 'SELECT requests, addon_requests, addon_r2_bytes, addon_d1_bytes FROM tenant_usage WHERE month = ? ORDER BY requests DESC LIMIT 1',
        args: [currentMonth()],
      })),
      meta: [],
    });

    let usage = { requests: 0, addon_requests: 0, addon_r2_bytes: 0, addon_d1_bytes: 0 };
    if (result.action === 'respond' && result.response) {
      const records = JSON.parse(new TextDecoder().decode(result.response.data));
      if (Array.isArray(records) && records.length > 0) {
        const r = records[0].data || records[0];
        usage = {
          requests: r.requests ?? 0,
          addon_requests: r.addon_requests ?? 0,
          addon_r2_bytes: r.addon_r2_bytes ?? 0,
          addon_d1_bytes: r.addon_d1_bytes ?? 0,
        };
      }
    }

    // Get storage usage
    const storageResult = await host.callBlock('wafer-run/database', {
      kind: 'database.query_raw',
      data: new TextEncoder().encode(JSON.stringify({
        query: 'SELECT COALESCE(SUM(size), 0) as total FROM storage_objects',
        args: [],
      })),
      meta: [],
    });

    let storageBytes = 0;
    if (storageResult.action === 'respond' && storageResult.response) {
      const records = JSON.parse(new TextDecoder().decode(storageResult.response.data));
      if (Array.isArray(records) && records.length > 0) {
        storageBytes = (records[0].data || records[0]).total ?? 0;
      }
    }

    return jsonRespond(msg, {
      month: currentMonth(),
      requests: {
        used: usage.requests,
        addon: usage.addon_requests,
      },
      storage: {
        r2_bytes: storageBytes,
        r2_addon_bytes: usage.addon_r2_bytes,
        d1_addon_bytes: usage.addon_d1_bytes,
      },
    });
  } catch {
    return jsonRespond(msg, { error: 'failed to fetch usage' });
  }
}

function currentMonth(): string {
  const d = new Date();
  return `${d.getUTCFullYear()}-${String(d.getUTCMonth() + 1).padStart(2, '0')}`;
}

function jsonRespond(msg: Message, data: unknown): BlockResult {
  return {
    action: 'respond',
    response: {
      data: new TextEncoder().encode(JSON.stringify(data)),
      meta: [{ key: 'resp.content_type', value: 'application/json' }],
    },
    message: msg,
  };
}
